exla/lib/exla/backend.ex

```diff
@@ -0,0 +1,268 @@
+defmodule EXLA.Backend do
+  @moduledoc """
+  A Nx tensor backend for the data kept on the device.
+
+  You can directly transfer to this backend by calling
+  `Nx.backend_transfer/2` or `Nx.backend_copy/2`. It
+  allows the following options:
+
+    * `:client` - the client to store the data on.
+      Defaults to the client configured in `Nx.Defn`,
+      otherwise uses `:host`.
+
+    * `:device_id` - which device to store it on
+
+  To get the data out of the device backend into a regular
+  tensor, call `Nx.backend_transfer/1` (with the device
+  tensor as the single argument).
+  """
+
+  @behaviour Nx.Backend
+  @enforce_keys [:buffer]
+  defstruct [:buffer]
+
+  alias Nx.Tensor, as: T
+  alias EXLA.Backend, as: B
+
+  import Nx.Shared
+
+  @impl true
+  def constant(%{type: type, shape: shape} = out, constant, backend_options) do
+    data = :binary.copy(number_to_binary(constant, type), Nx.size(shape))
+    from_binary(out, data, backend_options)
+  end
+
+  defp number_to_binary(number, type),
+    do: match_types([type], do: <<write!(number, 0)>>)
+
+  @impl true
+  def from_binary(%T{shape: shape, type: type} = tensor, binary, opts) do
+    {client, device_id} = client_and_device_id(opts)
+    shape = EXLA.Shape.make_shape(type, shape)
+    buffer = EXLA.DeviceBuffer.place_on_device(binary, shape, client, device_id)
+    put_in(tensor.data, %B{buffer: buffer})
+  end
+
+  @impl true
+  def backend_copy(tensor, Nx.Tensor, opts) do
+    backend_copy(tensor, Nx.BinaryBackend, opts)
+  end
+
+  # TODO: Support direct transfers without going through Elixir
+  def backend_copy(%T{data: %B{buffer: buffer}} = tensor, backend, opts) do
+    backend.from_binary(tensor, EXLA.DeviceBuffer.read(buffer), opts)
+  end
+
+  @impl true
+  def backend_transfer(%T{data: %B{buffer: buffer}} = tensor, backend, opts) do
+    if backend == __MODULE__ and same_client_device?(buffer, opts) do
+      tensor
+    else
+      try do
+        backend_copy(tensor, backend, opts)
+      after
+        EXLA.DeviceBuffer.deallocate(buffer)
+      end
+    end
+  end
+
+  @impl true
+  def backend_deallocate(%T{data: %B{buffer: buffer}}) do
+    EXLA.DeviceBuffer.deallocate(buffer)
+  end
+
+  @impl true
+  def to_batched_list(out, tensor, opts) do
+    binary_tensor = backend_copy(tensor, Nx.BinaryBackend, [])
+    Nx.BinaryBackend.to_batched_list(out, binary_tensor, opts)
+  end
+
+  @impl true
+  def to_binary(%T{data: %B{buffer: buffer}, type: {_, size}}, limit) do
+    EXLA.DeviceBuffer.read(buffer, limit * div(size, 8))
+  end
+
+  @impl true
+  def inspect(%T{} = tensor, inspect_opts) do
```

Comment:
Please update to the inspect implementation of this PR: https://github.com/elixir-nx/nx/pull/740 :D

Reply:
Ohhh, I mirrored the local version of that, thanks! :see_no_evil:

Reply:
@josevalim in `Nx.Backend.inspect/2` the docs say

https://github.com/elixir-nx/nx/blob/bd4c499f95e23ae7d76bee08b5f678c719a6644b/nx/lib/nx/backend.ex#L148-L150

Should we `+ 1` or is it no longer relevant?

Reply:
Good find. We need to +1.